### PR TITLE
fix(pol): enforce non-zero total stake

### LINKED ISSUE #2166: Make total_stake NonZero
For PoL, total stake must be >0

https://www.notion.so/nomos-tech/v1-1-Proof-of-Leadership-Specification-2e9261aa09df80058244c902defc6da2?source=copy_link#2e9261aa09df8010b506cea919608b8b

Change its type from `u64` to `NonZeroU64` everywhere. Also, make sure that the algorithm introduced in https://github.com/logos-blockchain/logos-blockchain/pull/2076 doesn't make total stake 0.

### PR DESCRIPTION
**Pull request by PoulavBhowmick03**
## 1. What does this PR implement?
Closes #2166 

## 2. Does the code have enough context to be clearly understood?

Yes

## 3. Who are the specification authors and who is accountable for this PR?

@youngjoon-lee , @danielSanchezQ 
PR: @PoulavBhowmick03

## 4. Is the specification accurate and complete?

Based on #2166 and the linked PoL section, the core requirement is accurate: PoL must not operate with `total_stake == 0`.

## 5. Does the implementation introduce changes in the specification?

No

## Checklist

> [!WARNING]  
> Do not merge the PR if any of the following is missing:

* [X] 1. The PR title follows the Conventional Commits [specification](https://www.notion.so/How-to-open-a-Pull-Request-215261aa09df80deb538ed59f5e4e0b5).
* [X] 2. Description added.
* [X] 3. Context and links to Specification document(s) added.
* [X] 4. Main contact(s) (developers and specification authors) added
* [ ] 5. Implementation and Specification are 100% in sync including changes. This is critical.
* [ ] 6. Link PR to a specific milestone.

### CODE REVIEW

File: blend/proofs/src/quota/fixtures.rs [hunk 1]
```diff
@@ -23,7 +25,7 @@ pub fn valid_proof_of_core_quota_inputs(
 ) -> (PublicInputs, ProofOfCoreQuotaInputs) {
     let (lottery_0, lottery_1) =
         LotteryConstants::new(NonNegativeRatio::new(1, 10.try_into().unwrap()))
-            .compute_lottery_values(1);
+            .compute_lottery_values(NonZeroU64::MIN);
```
Comment (ntn-x2) [inline blend/proofs/src/quota/fixtures.rs]: This is good, but it would be more explicit to use `1.try_into().unwrap()` (or any other way to initialize a `NonZeroU64`) here, so we know what value we are testing with.

File: core/src/proofs/leader_proof.rs [hunk 2]
```diff
@@ -358,7 +358,7 @@ mod tests {
     }
 
     fn rand_inputs(lottery_constants: &LotteryConstants) -> (LeaderPublic, Fr, Fr) {
-        let (lottery_0, lottery_1) = lottery_constants.compute_lottery_values(1);
+        let (lottery_0, lottery_1) = lottery_constants.compute_lottery_values(NonZeroU64::MIN);
```
Comment (ntn-x2) [inline core/src/proofs/leader_proof.rs]: Same as above.

File: services/blend/src/edge/mod.rs [hunk 3]
```diff
@@ -380,41 +380,78 @@ where
         PolEpochInfo,
         MessageHandler<Backend, NodeId, ProofsGenerator, RuntimeServiceId>,
     )> = None;
+    let mut session_stream_open = true;
```
Comment (ntn-x2) [inline services/blend/src/edge/mod.rs]: I guess this needs rebasing?

**Review by ntn-x2 — COMMENTED** [review]
Nice! Just a few nitpicks on my side. Thank you!

File: blend/proofs/src/quota/fixtures.rs (hunk 1 above)
Comment (PoulavBhowmick03) [reply blend/proofs/src/quota/fixtures.rs]: made the changes!

**Review by davidrusu — APPROVED** [review]
Looks good, thank you!

**Comment by davidrusu** [comment]
@PoulavBhowmick03 there are some cargo deny failures, this may have already been fixed in master, can you rebase?

File: services/chain/chain-network/src/bootstrap/ibd.rs [hunk 4]
```diff
@@ -1118,6 +1120,19 @@ mod tests {
         )
     }
 
+    fn genesis_utxo() -> lb_core::mantle::ledger::Utxo {
```
Comment (youngjoon-lee) [inline services/chain/chain-network/src/bootstrap/ibd.rs]: Could you please add a doc comment: "Chain should start with total_stake>=1"?

File: ledger/src/cryptarchia/mod.rs [hunk 5]
```diff
@@ -555,13 +558,14 @@ impl LedgerState {
...
+        let total_stake_raw = utxos
             .utxos()
             .iter()
             .filter(|(_, (utxo, _))| config.faucet_pk.is_none_or(|fpk| utxo.note.pk != fpk))
             .map(|(_, (utxo, _))| utxo.note.value)
-            .sum::<Value>()
-            .max(1); // TODO: Change total_stake to NonZeroU64: https://github.com/logos-blockchain/logos-blockchain/issues/2166
+            .sum::<Value>();
+        let total_stake = NonZeroU64::new(total_stake_raw)
+            .expect("genesis must have at least one non-faucet UTXO with positive stake");
```
Comment (youngjoon-lee) [inline ledger/src/cryptarchia/mod.rs:647]: A nice edge case handling 👍

**Review by youngjoon-lee — APPROVED** [review]
Great! Thank you.

File: services/chain/chain-network/src/bootstrap/ibd.rs [hunk 6]
```diff
@@ -1118,6 +1120,20 @@ mod tests {
         )
     }
 
+    /// Chain should start with `total_stake` ≥ 1
+    fn genesis_utxo() -> lb_core::mantle::ledger::Utxo {
```
Comment (danielSanchezQ) [inline services/chain/chain-network/src/bootstrap/ibd.rs]: nitpick: Remove the fully qualified names and use the type directly better.

File: services/chain/chain-network/src/bootstrap/ibd.rs [hunk 7]
```diff
@@ -1118,6 +1120,20 @@ mod tests {
...
+        use lb_core::mantle::ledger::{Inputs, Outputs};
```
Comment (danielSanchezQ) [inline services/chain/chain-network/src/bootstrap/ibd.rs]: nitpick: Why local import?

File: services/chain/chain-service/src/states.rs [hunk 8]
```diff
@@ -387,4 +392,17 @@ mod tests {
...
+        let zk_sk = ZkKey::from(BigUint::from(1u64));
+        let transfer = TransferOp::new(
+            Inputs::new(vec![]),
+            Outputs::new(vec![Note::new(1, zk_sk.to_public_key())]),
+        );
+        transfer
+            .outputs
+            .utxo_by_index(0, &transfer)
+            .expect("genesis transfer should create one UTXO")
+    }
```
Comment (danielSanchezQ) [inline services/chain/chain-service/src/states.rs]: This is duplicated, we can probably reuse a single function somehow.

**Review by danielSanchezQ — CHANGES_REQUESTED** [review]
Overall ok, just some cleanup comments.

**Comment by davidrusu** [comment]
@PoulavBhowmick03 checking in if you still intend to finish this.

**Comment by PoulavBhowmick03** [comment]
> @PoulavBhowmick03 checking in if you still intend to finish this.

Hey, apologies for this delay. I was having my exams, so was unable to look into this. I still intend to finish this. Will address the changes by today

**Comment by PoulavBhowmick03** [comment]
@davidrusu do take a look, thanks!

**Comment by danielSanchezQ** [comment]
@PoulavBhowmick03 this needs a big rebase/merge again if you dont mind. Thanks

**Comment by PoulavBhowmick03** [comment]
> @PoulavBhowmick03 this needs a big rebase/merge again if you dont mind. Thanks

hey yes, i am taking a look at this and getting it done

**Comment by PoulavBhowmick03** [comment]
@davidrusu @danielSanchezQ fixed the conflicts

**Comment by davidrusu** [comment]
@PoulavBhowmick03 the merge got completely botched, there is lots of code that is reappearing here that should not be.

**Comment by PoulavBhowmick03** [comment]
> @PoulavBhowmick03 the merge got completely botched, there is lots of code that is reappearing here that should not be.

Yeah I checked that, might have slipped as there were quite some conflicts. I am fixing these
